src/xml_document.cc

```diff
@@ -335,144 +350,112 @@ xmlParserOption getParserOptions(v8::Local<v8::Object> props) {
 
     ret |= getParserOption(props, "big_lines", XML_PARSE_BIG_LINES);        // 4194304: Store big lines numbers in text PSVI field
 
-    return (xmlParserOption)ret;
+    return ret;
 }
 
-NAN_METHOD(XmlDocument::FromHtml)
-{
-    Nan::HandleScope scope;
+xmlDoc* parseString(char* string, int length, char* baseUrl,
+                    char* encoding, int parserOptions, bool isHTML) {
+        if (isHTML) {
+            return htmlReadMemory(string, length, baseUrl, encoding, parserOptions);
+        }
+        else {
+            return xmlReadMemory(string, length, baseUrl, encoding, parserOptions);
+        }
+}
+
+xmlDoc* parseData(v8::Local<v8::Value> data, v8::Local<v8::Object> options, bool isHTML) {
+    char* baseUrl = getStringOption(
+                options->Get(Nan::New<v8::String>("baseUrl").ToLocalChecked()));
+
+    char* encoding = getStringOption(
+                options->Get(Nan::New<v8::String>("encoding").ToLocalChecked()));
+
+    int parserOptions = getParserOptions(options);
 
-    v8::Local<v8::Object> options = info[1]->ToObject();
-    v8::Local<v8::Value>  baseUrlOpt  = options->Get(
-        Nan::New<v8::String>("baseUrl").ToLocalChecked());
-    v8::Local<v8::Value>  encodingOpt = options->Get(
-        Nan::New<v8::String>("encoding").ToLocalChecked());
-    v8::Local<v8::Value> excludeImpliedElementsOpt = options->Get(
-        Nan::New<v8::String>("excludeImpliedElements").ToLocalChecked());
-
-    // the base URL that will be used for this HTML parsed document
-    v8::String::Utf8Value baseUrl_(baseUrlOpt->ToString());
-    const char * baseUrl = *baseUrl_;
-    if (!baseUrlOpt->IsString()) {
-        baseUrl = NULL;
+    if (node::Buffer::HasInstance(data)) {
+        v8::Local<v8::Object> buf = data->ToObject();
+        return parseString(node::Buffer::Data(buf),
+                            node::Buffer::Length(buf),
+                            baseUrl,
+                            encoding,
+                            parserOptions,
+                            isHTML);
+    }
+    else {
+        v8::String::Utf8Value str(data->ToString());
+        return parseString(*str,
+                            str.length(),
+                            baseUrl,
+                            encoding,
+                            parserOptions,
+                            isHTML);
     }
 
-    // the encoding to be used for this document
-    // (leave NULL for libxml to autodetect)
-    v8::String::Utf8Value encoding_(encodingOpt->ToString());
-    const char * encoding = *encoding_;
 
-    if (!encodingOpt->IsString()) {
-        encoding = NULL;
+    /*
+    xmlNode* root_node = xmlDocGetRootElement(doc);
+    if (root_node == NULL) {
+        return Nan::ThrowError("parsed document has no root element");
     }
+    */
```

Comment:
Yes I forgot to bring this up. I was going to ask why we were checking for the root node for fromXML and not fromHTML? Should we be checking for both? Or neither?